### PR TITLE
Diary: approximate time

### LINKED ISSUE #56202: The diary knows the precise time always
### Describe the bug

When creating a diary entry with no knowledge of the time, it will output the precise time to the second.

### Steps to reproduce

1. Remove all methods of timekeeping on you
2. Create diary entry
3. See exact time

### Expected behavior

The diary entry should either display "???", or basics such as Early Morning, Midday, Evening (if you are able to see the sky)

### Screenshots

_No response_

### Versions and configuration

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dseguin**
#### Summary
None

#### Purpose of change
- Fixes #56202

Adding an entry to the diary always displays the precise time even if the player has no timekeeping device.

#### Describe the solution
When adding a new diary entry:
- If the player has an active timekeeping device => **record the precise time**
- If the player doesn't have a timekeeping device...
  - If the player is outdoors (able to see the sky) => **record the approximate time of day**
  - If the player is inside/underground => **don't record the time at all**

Assuming the day of the current season can be tracked by the character on their own (if they can see the sky).

#### Describe alternatives you've considered
There are different ways to represent the approximate time, I just went with something close to how it's represented in the sidebar.

#### Testing
Precise time (using a watch):

precision_2

Approximate time (no watch, outdoors):

precision_1

No idea what time it is (no watch, underground):

precision_0

### CODE REVIEW

File: src/diary.cpp [hunk 1]
```diff
@@ -643,6 +710,7 @@ void diary::new_page()
     page -> kills = g ->get_kill_tracker().kills;
     page -> npc_kills = g->get_kill_tracker().npc_kills;
     avatar *u = &get_avatar();
+    page -> time_accuracy = u->has_watch() ? 2 : get_map().get_abs_sub().z() >= 0 ? 1 : 0;
```
Comment (Night-Pryanik) [inline src/diary.cpp:713]: Simply being above ground doesn't necessarily gives you ability to see the sky: character could be standing inside a building with closed doors and windows. I suggest using a check for `is_outside()` instead.

File: src/diary.h [hunk 2]
```diff
@@ -27,6 +27,8 @@ struct diary_page {
     std::vector<std::string> diff_to_previous_page;
     /*turn the page was created*/
     time_point turn;
+    /*accuracy of time recorded in journal entry*/
```
Comment (Night-Pryanik) [inline src/diary.h:30]:
```suggestion
    /*accuracy of time recorded in journal entry: 2 = player has a watch; 1 = player is able to see the sky; 0 = no idea what time is it now*/
```

File: src/diary.cpp [hunk 3]
```diff
@@ -579,39 +572,113 @@ std::string diary::get_page_text()
...
+            //~ Time of year: $1 = year since Cataclysm, $2 = season
+            return string_format( "Year %1$d, %2$s", year, season );
+        } else {
+            // partial accuracy, able to see the sky
+            //~ Time of year:
+            //~ $1 = year since Cataclysm
+            //~ $2 = season
+            //~ $3 = day of season
+            //~ $4 = approximate time of day
+            return string_format( "Year %1$d, %2$s, day %3$d, %4$s", year,
```
Comment (dseguin) [inline src/diary.cpp:648]:
```suggestion
            return string_format( _( "Year %1$d, %2$s, day %3$d, %4$s" ), year,
```

File: src/diary.cpp (hunk 1 above)
Comment (dseguin) [reply src/diary.cpp:713]: Sure, I can do that.

The "above ground" condition was just to match the way timekeeping works in the sidebar:

https://github.com/CleverRaven/Cataclysm-DDA/blob/f33592293975e6765c064c3f77a97fbccc75d7fc/src/panels.cpp#L570-L578

Comment (dseguin) [reply src/diary.cpp:713]:
```suggestion
    page -> time_accuracy = u->has_watch() ? 2 : is_creature_outside( *u ) ? 1 : 0;
```

Comment (Night-Pryanik) [reply src/diary.cpp]: > The "above ground" condition was just to match the way timekeeping works in the sidebar:

This was my own work from the #20953 :confused:

While you're here, you could replace sidebar timekeeping logic too.

File: src/diary.cpp [hunk 4]
```diff
@@ -579,39 +572,113 @@ std::string diary::get_page_text()
     return "";
 }
 
+static std::string get_time_since_str( const time_duration &turn_diff, int acc )
```
Comment (ZhilkinSerg) [inline src/diary.cpp:575]: This probably belongs to calendar.cpp.

File: src/diary.cpp [hunk 5]
```diff
@@ -781,6 +850,7 @@ void diary::deserialize( JsonIn &jsin )
             std::unique_ptr<diary_page> page( new diary_page() );
             page->m_text = elem.get_string( "text" );
             elem.read( "turn", page->turn );
+            elem.read( "time_acc", page->time_accuracy );
```
Comment (ZhilkinSerg) [inline src/diary.cpp:853]: Better use full string, i.e. `time_accuracy`.

File: src/diary.h [hunk 6]
```diff
@@ -27,6 +27,9 @@ struct diary_page {
     std::vector<std::string> diff_to_previous_page;
     /*turn the page was created*/
     time_point turn;
+    /*accuracy of time recorded in journal entry:
+    2 = player has a watch; 1 = player is able to see the sky; 0 = no idea what time is it now*/
+    int time_accuracy = 0;
```
Comment (ZhilkinSerg) [inline src/diary.h:32]: Should be `enum` instead of `int` to avoid confusion.

File: src/display.cpp [hunk 7]
```diff
@@ -206,7 +206,7 @@ std::string display::time_string( const Character &u )
     // Return exact time if character has a watch, or approximate time if aboveground
```
Comment (Night-Pryanik) [inline src/display.cpp:206]:
```suggestion
    // Return exact time if character has a watch, or approximate time if can see the sky
```

**Pull request merged**